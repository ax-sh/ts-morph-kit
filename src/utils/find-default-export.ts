import type { SourceFile } from "ts-morph";

export function findDefaultExport(sourceFile: SourceFile) {
  // Find the export default declaration
  const exportDefault = sourceFile
    .getExportAssignments()
    .find((assignment) => !assignment.isExportEquals());
  if (!exportDefault) {
    throw new Error("No 'export default' found in the source file.");
  }
  // console.log("Found 'export default':");
  const exportedExpression = exportDefault.getExpression();
  return exportedExpression;
}
