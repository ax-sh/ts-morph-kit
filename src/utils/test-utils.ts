import path from "node:path";
import { ScriptKind } from "ts-morph";
import { createProject } from "../core/create-project.ts";

export async function createMemorySourceFile(tempFile: string, input: string) {
  const project = createProject({ useInMemoryFileSystem: true });
  const sourceFile = project.createSourceFile(tempFile, input, {
    // Note: .js and .mjs can still be valid for TS projects.
    // We can't infer TypeScript from config.tsx.
    scriptKind:
      path.extname(tempFile) === ".ts" ? ScriptKind.TS : ScriptKind.JS,
  });

  return sourceFile;
}

export async function createTestSourceFile(code: string) {
  const resolvedPath = "test.ts";
  const sourceFile = await createMemorySourceFile(resolvedPath, code);
  return sourceFile;
}
