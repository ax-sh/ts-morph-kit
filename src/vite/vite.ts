import { findDefaultExport } from "../utils/find-default-export.ts";
import { CallExpression, type SourceFile, SyntaxKind } from "ts-morph";

export class NotAFunctionCallError extends Error {
  constructor() {
    super("The 'export default' is not a function call.");
    this.name = "NotAFunctionCallError";
  }
}

export class InvalidCalleeError extends Error {
  constructor(expectedFuncName: string) {
    super(`The 'export default' does not call '${expectedFuncName}'.`);
    this.name = "InvalidCalleeError";
  }
}

export class MissingArgumentsError extends Error {
  constructor(funcName: string) {
    super(`The '${funcName}' call has no arguments.`);
    this.name = "MissingArgumentsError";
  }
}

export class InvalidArgumentTypeError extends Error {
  constructor(funcName: string) {
    super(
      `The '${funcName}' call does not contain an object literal as its argument.`,
    );
    this.name = "InvalidArgumentTypeError";
  }
}

export function getDefaultViteConfig(sourceFile: SourceFile) {
  const exportedExpression = findDefaultExport(sourceFile);
  // Ensure the exported expression is a CallExpression
  if (!CallExpression.isCallExpression(exportedExpression)) {
    throw new NotAFunctionCallError();
  }
  // Check if the callee is `defineConfig`
  const callee = exportedExpression.getExpression();
  const funcName = "defineConfig";
  if (
    callee.getKind() !== SyntaxKind.Identifier ||
    callee.getText() !== funcName
  ) {
    throw new InvalidCalleeError(funcName);
  }
  // Get the arguments of the `defineConfig` call
  const args = exportedExpression.getArguments();
  if (args.length === 0) {
    throw new MissingArgumentsError(funcName);
  }

  const [configObject] = args;
  if (
    !configObject ||
    !configObject.isKind(SyntaxKind.ObjectLiteralExpression)
  ) {
    throw new InvalidArgumentTypeError(funcName);
  }
  return configObject;
}
