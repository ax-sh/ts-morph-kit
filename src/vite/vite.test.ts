import { findDefaultExport } from "../utils/find-default-export.ts"
import {
  formatSourceFile,
  formatSourceFileToString,
} from "../utils/format-source-file.ts"
import { createTestSourceFile } from "../utils/test-utils.ts"
// import { expect, describe, it } from "bun:test";
import {
  addBaseProperty,
  addVitePlugins,
  getDefaultViteConfig,
} from "./vite.ts"

describe("vitest config test", () => {
  const code = `
        import { defineConfig } from 'vite';
        // https://vite.dev/config/
        export default defineConfig({
        plugins:[foo()]
        base:"test"
        });
  ` as const
  it("should handle empty configuration objects", async () => {
    const code = `
      import { defineConfig } from 'vite';
      export default defineConfig({});
    `
    const sourceFile = await createTestSourceFile(code)
    const modifiedSourceFile = addBaseProperty(sourceFile, "/new-base")

    const result = formatSourceFileToString(modifiedSourceFile.getSourceFile())
    expect(result).toContain(`base: "/new-base"`)
    expect(result).toMatchSnapshot() // Optional: Use snapshots for regression testing
  })
  it("should load vite config", async () => {
    const sourceFile = await createTestSourceFile(code)
    const config = getDefaultViteConfig(sourceFile)
    expect(config).toBeDefined()
    console.log(config)
  })

  it("should ", async () => {
    const sourceFile = await createTestSourceFile(code)

    const exportedExpression = findDefaultExport(sourceFile)

    const functionName = exportedExpression.getText()
    const formatted = formatSourceFile(sourceFile).getFullText()
    console.log(formatted)
    expect(exportedExpression.getText()).toContain("defineConfig({")
    console.log("Exported Expression:")
    // console.log();
  })

  it("should modify plugins without duplicates", async () => {
    const sf = await createTestSourceFile(code)
    const modified = addVitePlugins(sf, ["dff()", "dff", "dff"])

    console.log(formatSourceFileToString(modified.getSourceFile()))
  })
  it("should modify addBaseProperty without duplicates", async () => {
    const sf = await createTestSourceFile(code)
    const modified = addBaseProperty(sf, "voo")

    console.log(formatSourceFileToString(modified.getSourceFile()))
  })

  it.todo("should handle missing `defineConfig` call", async () => {
    const code = `
      export default defineConfig({});
    `
    const sourceFile = await createTestSourceFile(code)

    expect(() => addBaseProperty(sourceFile, "/new-base")).toThrowError(
      "`plugins` property not found in the `defineConfig` object.",
    )
  })
})
