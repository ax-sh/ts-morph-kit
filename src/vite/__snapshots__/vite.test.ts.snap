// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`vitest config test > should handle empty configuration objects 1`] = `
"
import { defineConfig } from 'vite';
export default defineConfig({
  base: "/new-base"
});
"
`;

exports[`vitest config test should handle empty configuration objects 1`] = `
"
import { defineConfig } from 'vite';
export default defineConfig({
  base: "/new-base"
});
"
`;
